PICARD-1718: Try to log and show exception details on crash

phw opened the pull request:
<!--
    Hello! Thanks for submitting a pull request to MusicBrainz Picard. We
    appreciate your time and interest in helping our project!

    Use this template to help us review your change. Not everything is required,
    depending on your change. Keep or delete what is relevant for your change.
    Remember that it helps us review if you give more helpful info for us to
    understand your change.

    Ensure that you've read through and followed the Contributing Guidelines, in
    [CONTRIBUTING.md](https://github.com/metabrainz/picard/blob/master/CONTRIBUTING.md).
-->

# Summary
If Picard crashes try to log the traceback to a file and show a dialog to the user before exiting.
<!--
    Update the checkbox with an [x] for the type of contribution you are making.
-->

* This is a…
    * [ ] Bug fix
    * [x] Feature addition
    * [ ] Refactoring
    * [ ] Minor / simple change (like a typo)
    * [ ] Other
* **Describe this change in 1-2 sentences**:

# Problem
If Picard crashes due to some bug error output is written to stdout. But most GUI users won't see those messages unless they explicitly start Picard from command line. This has some issues:

1. The crash that just happened may not be easily reproducible. So even if the user starts from command line they will probably not trigger this issue.

2. For users of macOS and especially Windows this is more difficult. Especially for Windows users this is often unknown territory and the default terminal of Windows has pretty bad usability

3. On Windows the portable install and the Windows store version don't write to stdout. There is no convenient way to start the store version from command line at all.
<!--
    Anything that helps us understand why you are making this change goes here.
    What problem are you trying to fix? What does this change address?
-->

* JIRA ticket (_optional_): PICARD-1718
<!--
    Please make sure you prefix your pull request title with 'PICARD-XXX' in order
    for our ticket tracker to link your pull request to the relevant ticket.
-->



# Solution
Catch all exceptions and display a dialog box with traceback defaults. This works fine on typical Python issues, including syntax errors or crashes caused by plugins.

To discuss:

1. This won't fetch crashes caused by native code like memory access validations. I had some experiments with enabling the faulthandler module available since Python 3.3, see https://github.com/phw/picard/commit/0fa2aaac091108ebfaeb20f635b31626d63de083 . While I was able to fetch coredumps on Linux with this it did have no effect on Windows, which I kind of see the primary target for this.

2. I am annoyed by the code duplication for both `tagger.py` and `picard`. But I really want to keep this code inline. I think the actual goal should be to get rid of the duplicate entry points and use only a single one, but I did not want to start this change here.

<!--
    The details of your change. Talk about technical details, considerations, or
    other interesting points. If you have a lot to say, be more detailed in this
    section.
-->


# Action

<!--
    Other than merging your change, do you want / need us to do anything else
    with your change? This could include reviewing a specific part of your PR.
-->

zas commented on scripts/picard.in:
```diff
@@ -1,3 +1,30 @@
+        os.write(fd, trace.encode(errors="replace"))
+        os.close(fd)
+    except:  # noqa: F722
+        print("Failed writing log file {0}".format(logfile), file=sys.stderr)
```
In this case, the following message is incorrect (A log file hasn't been written)

zas commented on scripts/picard.in:
```diff
@@ -1,3 +1,30 @@
+    logfile = None
+    try:
+        (fd, logfile) = tempfile.mkstemp(".log", "picard-crash-")
+        os.write(fd, trace.encode(errors="replace"))
```
Is there a reason to not use `NamedTemporaryFile()` together with `with` here?

```python
with NamedTemporaryFile(suffix='.log', prefix='picard-crash-', delete=False) as f:
   f.write(trace.encode(errors="replace"))
   logfile = f.name
```

phw replied:
I mainly did this because the file must not be deleted. But I see know using NamedTemporaryFile with delete=False achieves this as well.

zas commented on tagger.py.in:
```diff
@@ -15,5 +15,29 @@ else:
+    # Display the crash information to the user as a dialog. This requires#
+    # importing Qt5 and has some potential to fail if things are broken.
+    from PyQt5 import QtWidgets
+    app = QtWidgets.QApplication(sys.argv)
```
`app` isn't used anywhere

phw replied:
Yes, but the variable is needed to keep the object alive. Just using `QtWidgets.QApplication(sys.argv)` without assignment will not work and crash the application.

zas replied:
In this case, may be name it `unused` or the like to make clear it is intentional.

phw commented:
> I'm a bit bothered by code duplication, betwen `picard.in` and `tagger.py.in`

See my comment above. Actually I'm more bothered that we have two entrypoints at all. We need to get rid of scripts/picard.in

zas commented on scripts/picard.in:
```diff
@@ -1,3 +1,33 @@
+    if logfile:
+        msg.append("A logfile has been written to {0}".format(logfile))
+    msg.append("Error details: \n{0}".format(trace))
+    QtWidgets.QMessageBox.critical(None, "Picard terminated unexpectedly", "\n\n".join(msg))
```
It seems HTML-like tags are interpreted by `QtWidgets.QMessageBox.critical()`. It is an issue since we can't control what's in the trace text, it also concerns file path (while I doubt it can contain characters triggering RichText autodetection, better be safe).

Unfortunately, you can't change text format (and autodetection) when using static QMessageBox methods.

I did some experiments, and this trace is a good canditate for detailed text feature of QMessageBox, it is in text plain format by default:

Here is the test I did:

```python
import sys
from PyQt5.QtCore import Qt
from PyQt5.QtWidgets import QApplication, QMessageBox

app = QApplication(sys.argv)

msgbox = QMessageBox()
msgbox.setIcon(QMessageBox.Critical)
msgbox.setWindowTitle("Picard crashed")
msgbox.setTextFormat(Qt.PlainText)
msg = ["Picard <b>crashed</b>"]
logfile = '/x/<b>dddd</b>'
if logfile:
    msg.append("A logfile has been written to {0}".format(logfile))
msgbox.setText("\n\n".join(msg))
msgbox.setDetailedText("This is the trace<b>xxx<b>\nsfklslk")
msgbox.setStandardButtons(QMessageBox.Close)
msgbox.setDefaultButton(QMessageBox.Close)
msgbox.exec()
```
I intentionnaly used HTML-like tags in texts to test whether autodetection was on and if output was actually in plain text.
Here is the result on my system:
![image](https://user-images.githubusercontent.com/151042/74205086-162b2680-4c76-11ea-9909-7b924b575ad9.png)

Of course, it isn't as simple as original code, but it's the only way to prevent text format autodetection.

phw commented:
@zas I addressed your concerns about the HTML formatting in https://github.com/metabrainz/picard/pull/1466/commits/04547437509c61193f1d72bc1df35a49461d698c . Actually if we need to go the more complex route we can make use of this.

But I will actually advise we don't merge this PR yet. It is a nice idea, but this simple approach is not working as it should. E.g. you won't get any crashes catched by this if they happen anywhere else then the main thread. E.g. throw an exception in the options dialog and Picard will still just crash.

This needs more work. Either we can actually try to utilize the faulthandler or maybe it needs an actual wrapper process.

zas commented:
> E.g. you won't get any crashes catched by this if they happen anywhere else then the main thread. E.g. throw an exception in the options dialog and Picard will still just crash.

Ok :(

mineo commented on scripts/picard.in:
```diff
@@ -1,3 +1,45 @@
+        msgbox.setInformativeText(
+            'A logfile has been written to <a href="{0}">{1}</a>.'
+            .format(logfile_url.url(), logfile))
+    msgbox.setDetailedText(trace)
```
I'd show the stack trace by default and include a strong hint to include it in the ticket. Not every user is computer-savvy enough to do that by default and the log file might not survive the reboot that's inevitably going to happen between opening the ticket and being asked to provide the traceback.

phw commented:
I close this for now, the results where not really satisfying. Should I take this up again I'll create a new PR.
